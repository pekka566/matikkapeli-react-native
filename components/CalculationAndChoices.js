//import liraries
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import Calculation from './Calculation';
import Choices from './Choices';
import PropTypes from 'prop-types';

const CalculationAndChoices = props => {
  const calculation = (
    <Calculation calculation={props.calculation} answer={props.answer} />
  );

  const choices = (
    <Choices
      choices={props.choices}
      handleAnswer={answer => props.handleAnswer(answer)}
      buttonsDisabled={props.buttonsDisabled}
    />
  );
  return (
    <View
      style={{
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'space-around'
      }}
    >
      {calculation}
      {choices}
    </View>
  );
};

CalculationAndChoices.propTypes = {
  calculation: PropTypes.object.isRequired,
  answer: PropTypes.object,
  choices: PropTypes.array.isRequired,
  handleAnswer: PropTypes.func.isRequired,
  buttonsDisabled: PropTypes.bool.isRequired
};

export default CalculationAndChoices;
