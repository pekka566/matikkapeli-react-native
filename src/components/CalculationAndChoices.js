//import liraries
import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import Calculation from './Calculation';
import Choices from './Choices';
import PropTypes from 'prop-types';

const CalculationAndChoices = props => {
  const calculation = (
    <Calculation calculation={props.calculation} answer={props.answer} />
  );

  const choices = (
    <Choices
      choices={props.choices}
      handleAnswer={answer => props.handleAnswer(answer)}
      buttonsDisabled={props.buttonsDisabled}
    />
  );
  return (
    <View
      style={{
        flex: 0.7,
        flexDirection: 'column',
        justifyContent: 'flex-start'
      }}
    >
      {calculation}
      {choices}
    </View>
  );
};

CalculationAndChoices.defaultProps = {
  buttonsDisabled: false
};

CalculationAndChoices.propTypes = {
  calculation: PropTypes.object.isRequired,
  answer: PropTypes.object,
  choices: PropTypes.array.isRequired,
  handleAnswer: PropTypes.func.isRequired,
  buttonsDisabled: PropTypes.bool
};

export default CalculationAndChoices;
