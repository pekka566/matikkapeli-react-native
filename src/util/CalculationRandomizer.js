export const createCalculation = level => {
  let min;
  let max;
  switch (level) {
    case 1:
      min = 1;
      max = 10;
      break;
    case 2:
      min = 1;
      max = 20;
      break;
    case 999:
      min = 0;
      max = 99999;
      break;
    default:
      min = 0;
      max = 100;
  }

  const value = Math.floor(Math.random() * 4) + 1;
  let operator = '+';
  if (level > 1) {
    switch (value) {
      case 2:
        operator = '-';
        break;
      case 3:
        operator = '*';
        break;
      case 4:
        operator = '/';
        break;
      default:
        operator = '+';
        break;
    }
  }

  let firstNumber = Math.floor(Math.random() * max) + min;
  const secondNumber = Math.floor(Math.random() * firstNumber) + min;

  let result = firstNumber + secondNumber;
  if (operator === '-') {
    result = firstNumber - secondNumber;
  } else if (operator === '*') {
    result = firstNumber * secondNumber;
  } else if (operator === '/') {
    firstNumber = firstNumber * secondNumber;
    result = firstNumber / secondNumber;
  }

  let calculation = {
    firstNumber,
    operation: operator,
    secondNumber,
    result
  };
  return calculation;
};

const shuffleArray = array => {
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
};

const createChoisesArray = (calculation, choices, numOfChoices) => {
  const result = calculation.result;
  if (choices.length === 0) {
    choices.push({ value: result, isCorrect: true });
  }
  if (choices.length === numOfChoices) {
    shuffleArray(choices);
    return;
  }

  const addOrMinusMax = numOfChoices * 2; // 8
  const random = Math.floor(Math.random() * addOrMinusMax) + 1; // 1 to 8
  const doAdd = Math.round(Math.random()) === 1; // 0 or 1 -> boolean
  const operator = calculation.operation;
  const secondNumber = doAdd
    ? calculation.secondNumber + random
    : calculation.secondNumber - random;
  const firstNumber = calculation.firstNumber;
  let value = firstNumber + secondNumber;
  if (operator === '-') {
    value = firstNumber - secondNumber;
  } else if (operator === '*') {
    value = firstNumber * secondNumber;
  } else if (operator === '/') {
    value = firstNumber / secondNumber;
  }
  value = Math.round(value);
  if (
    choices.find(element => {
      return value === element.value;
    }) ||
    value < 1 ||
    value === result
  ) {
    createChoisesArray(calculation, choices, numOfChoices);
  } else {
    choices.push({ value, isCorrect: false });
    createChoisesArray(calculation, choices, numOfChoices);
  }
};

export const createChoices = (calculation, level) => {
  let choicesArr = [];
  createChoisesArray(calculation, choicesArr, 4);
  return choicesArr;
};
