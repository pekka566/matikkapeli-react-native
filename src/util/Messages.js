export const Messages = {
  en: {
    gamename: 'Math game',
    Level: 'Level',
    Select_level: 'Select level',
    hard: 'hard',
    Points: 'Points',
    Level_completed: 'Level completed!',
    Correct_answers: 'Correct answers',
    Got_points: 'You got {points} points.',
    Time_used: 'Time used {seconds} seconds.',
    Restart_level: 'Restart level',
    Restart_game: 'Restart game'
  },
  fi: {
    gamename: 'Matikkapeli',
    Level: 'Taso',
    Select_level: 'Valitse taso',
    hard: 'vaikea',
    Points: 'Pisteet',
    Level_completed: 'Taso suoritettu!',
    Correct_answers: 'Oikeita vastauksia',
    Got_points: 'Sait {points} pistett√§.',
    Time_used: 'Aikaa kului {seconds} sekuntia.',
    Restart_level: 'Aloita taso uudelleen',
    Restart_game: 'Aloita peli uudelleen'
  }
};
