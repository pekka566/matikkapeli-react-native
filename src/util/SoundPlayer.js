import Sound from 'react-native-sound';

// Enable playback in silence mode (iOS only)
Sound.setCategory('Playback', false); // true to mixWithOthers

let playingSounds = [];
const sounds = [
  { filename: 'game_sound_wrong.mp3' },
  { filename: 'game_sound_correct.mp3' }
];

class SoundPlayer {
  static stopCurrentPlayingSounds() {
    for (let sound of playingSounds) {
      sound.stop(() => {
        sound.release();
      });
    }
    playingSounds = [];
  }

  static playWrongSound(callback) {
    return SoundPlayer.play(0, callback);
  }

  static playCorrectSound(callback) {
    return SoundPlayer.play(1, callback);
  }

  static play(soundKey, callback, shouldStopOther = true) {
    if (shouldStopOther) {
      SoundPlayer.stopCurrentPlayingSounds();
    }

    return new Promise((resolve, reject) => {
      let data = sounds[soundKey];
      let sound = new Sound(data.filename, Sound.MAIN_BUNDLE, error => {
        if (error) {
          console.warn('Failed to load', data, error);
          reject(new Error('Failed to load ' + data.filename));
          return;
        }

        playingSounds.push(sound);
        sound.play(success => {
          // Release when it's done so we're not using up resources
          sound.release();

          if (success) {
            resolve(data);
            callback();
            return;
          }

          // reset the player to its uninitialized state (android only)
          // this is the only option to recover after an error occured and use the player again
          sound.reset();
          // failed due to audio decoding errors
          console.warn('Failed to play', soundKey, data.filename, sound._key);
          reject(new Error('Failed to play ' + data.filename));
        });
      }); // new Sound
    }); // new Promise
  }
}

export default SoundPlayer;
