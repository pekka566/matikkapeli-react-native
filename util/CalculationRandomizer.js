// If level = 1 return numbers between 1 to 10.

export const createCalculation = level => {
  let min;
  let max;
  switch (level) {
    case 1:
      min = 1;
      max = 10;
      break;
    default:
      min = 0;
      max = 100;
  }

  const firstNumber = Math.floor(Math.random() * max) + min;
  const secondNumber = Math.floor(Math.random() * max) + min;
  const result = firstNumber + secondNumber;

  let calculation = {
    firstNumber,
    operation: '+',
    secondNumber,
    result
  };
  return calculation;
};

const shuffleArray = array => {
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
};

const createChoisesArray = (result, choices, numOfChoices) => {
  if (choices.length === 0) {
    choices.push({ value: result, isCorrect: true });
  }
  if (choices.length === numOfChoices) {
    shuffleArray(choices);
    return;
  }

  const value = Math.floor(Math.random() * 20) + 1;
  if (
    choices.find(element => {
      return value === element.value;
    }) ||
    value < 1 ||
    value === result
  ) {
    createChoisesArray(result, choices, numOfChoices);
  } else {
    choices.push({ value, isCorrect: false });
    createChoisesArray(result, choices, numOfChoices);
  }
};

export const createChoices = result => {
  let choicesArr = [];
  createChoisesArray(result, choicesArr, 4);
  return choicesArr;
};
