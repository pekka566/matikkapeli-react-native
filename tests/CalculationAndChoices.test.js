import 'react-native';
import React from 'react';
import CalculationAndChoices from '../src/components/CalculationAndChoices';
import { createChoices } from '../src/util/CalculationRandomizer';

import renderer from 'react-test-renderer';

const calculation = {
  firstNumber: 5,
  operation: '+',
  secondNumber: 6,
  result: 11
};

const choices = [
  { value: 14, isCorrect: false },
  { value: 18, isCorrect: false },
  { value: 11, isCorrect: true },
  { value: 17, isCorrect: false }
];
const handlePressMock = jest.fn();

it('renders correctly', () => {
  const tree = renderer
    .create(
      <CalculationAndChoices
        calculation={calculation}
        choices={choices}
        handleAnswer={handlePressMock}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
