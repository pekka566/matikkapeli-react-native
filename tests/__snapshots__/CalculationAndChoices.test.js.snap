// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 0.7,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 0.1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginTop": "1%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "rgb(50, 50, 50)",
          "flex": 1,
          "fontSize": 30,
          "fontWeight": "bold",
          "textAlign": "center",
          "textShadowOffset": Object {
            "height": 1,
            "width": 1,
          },
        }
      }
    >
      5 
      +
       6 = 
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-around",
        "paddingBottom": "5%",
        "paddingTop": "5%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": "2%",
          "marginRight": "2%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#9b59b6",
            "borderColor": "#333",
            "borderRadius": 10,
            "borderWidth": 3,
            "flexDirection": "row",
            "height": 44,
            "justifyContent": "center",
            "marginBottom": 10,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "flex": 1,
                "fontSize": 30,
                "fontWeight": "bold",
                "lineHeight": 100,
                "textAlign": "center",
                "textShadowOffset": Object {
                  "height": 5,
                  "width": 5,
                },
                "textShadowRadius": 6,
              }
            }
          >
            14
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": "2%",
          "marginRight": "2%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#9b59b6",
            "borderColor": "#333",
            "borderRadius": 10,
            "borderWidth": 3,
            "flexDirection": "row",
            "height": 44,
            "justifyContent": "center",
            "marginBottom": 10,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "flex": 1,
                "fontSize": 30,
                "fontWeight": "bold",
                "lineHeight": 100,
                "textAlign": "center",
                "textShadowOffset": Object {
                  "height": 5,
                  "width": 5,
                },
                "textShadowRadius": 6,
              }
            }
          >
            18
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": "2%",
          "marginRight": "2%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#9b59b6",
            "borderColor": "#333",
            "borderRadius": 10,
            "borderWidth": 3,
            "flexDirection": "row",
            "height": 44,
            "justifyContent": "center",
            "marginBottom": 10,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "flex": 1,
                "fontSize": 30,
                "fontWeight": "bold",
                "lineHeight": 100,
                "textAlign": "center",
                "textShadowOffset": Object {
                  "height": 5,
                  "width": 5,
                },
                "textShadowRadius": 6,
              }
            }
          >
            11
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": "2%",
          "marginRight": "2%",
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#9b59b6",
            "borderColor": "#333",
            "borderRadius": 10,
            "borderWidth": 3,
            "flexDirection": "row",
            "height": 44,
            "justifyContent": "center",
            "marginBottom": 10,
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "flex": 1,
                "fontSize": 30,
                "fontWeight": "bold",
                "lineHeight": 100,
                "textAlign": "center",
                "textShadowOffset": Object {
                  "height": 5,
                  "width": 5,
                },
                "textShadowRadius": 6,
              }
            }
          >
            17
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
